### PR TITLE
[clang] solve crash due to function overloading.

### LINKED ISSUE #88917: Segfault when compiling weird code
The following file causes a segfault when compiled with clang 18.1.1:

```c
int b();
int main() { return b(b); }
int b(int (*f)()){
  return 0;
}
```

```
clang.c:2:22: warning: passing arguments to 'b' without a prototype is deprecated in all versions of C and is not supported in C23 [-Wdeprecated-non-prototype]
    2 | int main() { return b(b); }
      |                      ^
clang.c:1:5: warning: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C23, conflicting with a subsequent definition [-Wdeprecated-non-prototype]
    1 | int b();
      |     ^
clang.c:3:5: note: conflicting prototype is here
    3 | int b(int (*f)()){
      |     ^
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/bin/clang-18 -cc1 -triple x86_64-redhat-linux-gnu -emit-obj -mrelax-all -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name clang.c -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/tmp -fcoverage-compilation-dir=/tmp -resource-dir /usr/bin/../lib/clang/18 -internal-isystem /usr/bin/../lib/clang/18/include -internal-isystem /usr/local/include -internal-isystem /usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../x86_64-redhat-linux/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -ferror-limit 19 -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/clang-d07602.o -x c clang.c
1.      <eof> parser at end of file
2.      clang.c:3:5: LLVM IR generation of declaration 'b'
3.      clang.c:3:5: Generating code for declaration 'b'
 #0 0x00007f398960440a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM.so.18.1+0x60440a)
 #1 0x00007f3989601bf4 llvm::sys::RunSignalHandlers() (/lib64/libLLVM.so.18.1+0x601bf4)
 #2 0x00007f3989604b6b (/lib64/libLLVM.so.18.1+0x604b6b)
 #3 0x00007f3988a53710 __restore_rt (/lib64/libc.so.6+0x40710)
 #4 0x00007f3991e9cdcc (/lib64/libclang-cpp.so.18.1+0x1c9cdcc)
 #5 0x00007f3991e8a866 clang::CodeGen::CodeGenModule::GetOrCreateLLVMFunction(llvm::StringRef, llvm::Type*, clang::GlobalDecl, bool, bool, bool, llvm::AttributeList, clang::CodeGen::ForDefinition_t) (/lib64/libclang-cpp.so.18.1+0x1c8a866)
 #6 0x00007f3991e9175a clang::CodeGen::CodeGenModule::GetAddrOfFunction(clang::GlobalDecl, llvm::Type*, bool, bool, clang::CodeGen::ForDefinition_t) (/lib64/libclang-cpp.so.18.1+0x1c9175a)
 #7 0x00007f3991e8f5be clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/lib64/libclang-cpp.so.18.1+0x1c8f5be)
 #8 0x00007f3991e87975 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/lib64/libclang-cpp.so.18.1+0x1c87975)
 #9 0x00007f3991e8c38d clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (/lib64/libclang-cpp.so.18.1+0x1c8c38d)
#10 0x00007f3991e864c6 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (/lib64/libclang-cpp.so.18.1+0x1c864c6)
#11 0x00007f3991f3a85c (/lib64/libclang-cpp.so.18.1+0x1d3a85c)
#12 0x00007f3991e59042 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/lib64/libclang-cpp.so.18.1+0x1c59042)
#13 0x00007f399077444a clang::ParseAST(clang::Sema&, bool, bool) (/lib64/libclang-cpp.so.18.1+0x57444a)
#14 0x00007f3992a1c896 clang::FrontendAction::Execute() (/lib64/libclang-cpp.so.18.1+0x281c896)
#15 0x00007f39929945a0 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib64/libclang-cpp.so.18.1+0x27945a0)
#16 0x00007f3992a99df5 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib64/libclang-cpp.so.18.1+0x2899df5)
#17 0x00000000004144c6 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang-18+0x4144c6)
#18 0x0000000000411164 (/usr/bin/clang-18+0x411164)
#19 0x000000000040fea8 clang_main(int, char**, llvm::ToolContext const&) (/usr/bin/clang-18+0x40fea8)
#20 0x000000000041dcba main (/usr/bin/clang-18+0x41dcba)
#21 0x00007f3988a3d088 __libc_start_call_main (/lib64/libc.so.6+0x2a088)
#22 0x00007f3988a3d14b __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2a14b)
#23 0x000000000040d8a5 _start (/usr/bin/clang-18+0x40d8a5)
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 18.1.1 (Fedora 18.1.1-1.fc40)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Configuration file: /etc/clang/clang.cfg
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/clang-c82ad8.c
clang: note: diagnostic msg: /tmp/clang-c82ad8.sh
clang: note: diagnostic msg: 

********************
```

`/tmp/clang-c82ad8.c`:

```c
# 1 "<built-in>"
# 1 "clang.c"
int b();
int main() { return b(b); }
int b(int (*f)()){
  return 0;
}
```

`/tmp/clang-c82ad8.sh`:

```sh
# Crash reproducer for clang version 18.1.1 (Fedora 18.1.1-1.fc40)
# Driver args: "--gcc-triple=x86_64-redhat-linux" "clang.c"
# Original command:  "/usr/bin/clang-18" "-cc1" "-triple" "x86_64-redhat-linux-gnu" "-emit-obj" "-mrelax-all" "-dumpdir" "a-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "clang.c" "-mrelocation-model" "static" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/tmp" "-fcoverage-compilation-dir=/tmp" "-resource-dir" "/usr/bin/../lib/clang/18" "-internal-isystem" "/usr/bin/../lib/clang/18/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../x86_64-redhat-linux/include" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fskip-odr-check-in-gmf" "-fcolor-diagnostics" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "/tmp/clang-d07602.o" "-x" "c" "clang.c"
 "/usr/bin/clang-18" "-cc1" "-triple" "x86_64-redhat-linux-gnu" "-emit-obj" "-mrelax-all" "-dumpdir" "a-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "clang.c" "-mrelocation-model" "static" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/tmp" "-fcoverage-compilation-dir=/tmp" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fskip-odr-check-in-gmf" "-fcolor-diagnostics" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c" "clang-c82ad8.c"
```

### PR DESCRIPTION
**Pull request by lolloz98**
Closes #88917 

I am a newby in llvm development. 

I followed the suggestion for solving the issue. I added just a simple test to verify that the compilation does not crash.

Should the test not be good enough or any problems are found, please let me know, eager to learn :)

### CODE REVIEW

File: clang/test/CodeGen/function_overload_crash.c [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang_cc1 -emit-llvm < %s
```
Comment (efriedma-quic) [inline clang/test/CodeGen/function_overload_crash.c:1]: Can you integrate this test into some existing file that's testing something similar?

Please leave the bug number in a comment: something like `//GH88917`.

Please use FileCheck CHECK lines to verify the correct call instruction is emitted.

Comment (lolloz98) [reply clang/test/CodeGen/function_overload_crash.c:1]: Hello, I apologize, but could I ask some pointers to some docs for creating tests for llvm? I didn't find any very good resource about it :)

Comment (efriedma-quic) [reply clang/test/CodeGen/function_overload_crash.c:1]: Have you seen https://llvm.org/docs/TestingGuide.html#regression-test-structure ?

If you have any ideas for improving the documentation, please let me know (or better, propose a patch; it's generated from llvm/docs/TestingGuide.rst).

Comment (lolloz98) [reply clang/test/CodeGen/function_overload_crash.c:1]: Yes. I gave another read now and found out the FileCheck guide to solve some of my doubts. 
I will need to double check also the llvm IR generation, if everything is ok, I should be able to modify the test quickly.
As of now I am for a bit more than a week away from pc, but as soon as I am back I will finish this pr.

thank you for your help, I'll be posting soon :)

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by efriedma-quic** [comment]
Bot is unreliable, so copy-pasting:

> ⚠️ We detected that you are using a GitHub private e-mail address to contribute to the repo.
Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account.
See [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

**Pull request merged**